File: app/src/main/java/com/bopr/android/smailer/Notifications.java
package com.bopr.android.smailer;

import android.app.Activity;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.app.NotificationCompat.Builder;
import androidx.core.app.TaskStackBuilder;

import com.bopr.android.smailer.ui.HistoryActivity;
import com.bopr.android.smailer.ui.MainActivity;
import com.bopr.android.smailer.ui.RecipientsActivity;
import com.bopr.android.smailer.ui.RemoteControlActivity;
import com.bopr.android.smailer.ui.RulesActivity;
import com.bopr.android.smailer.util.TagFormatter;

import java.lang.annotation.Retention;

import static android.app.Notification.CATEGORY_ERROR;
import static android.app.Notification.CATEGORY_MESSAGE;
import static android.app.Notification.CATEGORY_SERVICE;
import static java.lang.annotation.RetentionPolicy.SOURCE;

/**
 * Produces notifications.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
@SuppressWarnings("WeakerAccess")
public class Notifications {

    private static final String CHANNEL_ID = "com.bopr.android.smailer";

    @IntDef({TARGET_MAIN, TARGET_RULES, TARGET_HISTORY, TARGET_RECIPIENTS, TARGET_REMOTE_CONTROL})
    @Retention(SOURCE)
    public @interface Target {
    }

    public static final int TARGET_MAIN = 0;
    public static final int TARGET_RULES = 1;
    public static final int TARGET_HISTORY = 2;
    public static final int TARGET_RECIPIENTS = 3;
    public static final int TARGET_REMOTE_CONTROL = 4;

    private static int messageId = -1;
    private static int errorId = -1;

    private Context context;
    private NotificationManager manager;
    private TagFormatter formatter;

    public Notifications(Context context) {
        this.context = context;
        manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        formatter = new TagFormatter(context);
    }

    public Notification getServiceNotification() {
        Builder builder = builder(context.getString(R.string.service_running), null, TARGET_MAIN);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setCategory(CATEGORY_SERVICE);
        }

        return builder.build();
    }

    public void showMessage(@StringRes int messageRes, @Target int target) {
        showMessage(context.getString(messageRes), target);
    }

    public void showError(@StringRes int messageRes, @Target int target) {
        showError(context.getString(messageRes), target);
    }

    public void showMailError(@StringRes int reasonRes, @Target int target) {
        showError(formatter
                        .pattern(R.string.unable_send_email)
                        .put("reason", reasonRes)
                        .format(),
                target);
    }

    public void showRemoteAction(@StringRes int messageRes, String argument) {
        showMessage(formatter
                        .pattern(messageRes)
                        .put("text", argument)
                        .format(),
                TARGET_HISTORY);
    }

    public void hideAllErrors() {
        while (errorId >= 0) {
            manager.cancel("error", errorId--);
        }
    }

    private void showMessage(String text, @Target int target) {
        Builder builder = builder(text, null, target).setAutoCancel(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setCategory(CATEGORY_MESSAGE);
        }

        manager.notify("message", ++messageId, builder.build());
    }

    private void showError(String text, @Target int target) {
        Builder builder = builder(text, context.getString(R.string.tap_to_check_settings), target).setAutoCancel(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setCategory(CATEGORY_ERROR);
        }

        manager.notify("error", ++errorId, builder.build());
    }

    private Builder builder(@NonNull String title, @Nullable String text, @Target int target) {
        return new Builder(context, getChannel())
                .setContentIntent(createIntent(target))
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(title)
                .setContentText(text);
    }

    @Nullable
    private PendingIntent createIntent(@Target int target) {
        switch (target) {
            case TARGET_MAIN:
                return createActivityIntent(MainActivity.class);
            case TARGET_HISTORY:
                return createActivityIntent(HistoryActivity.class);
            case TARGET_RECIPIENTS:
                return createActivityIntent(RecipientsActivity.class);
            case TARGET_REMOTE_CONTROL:
                return createActivityIntent(RemoteControlActivity.class);
            case TARGET_RULES:
                return createActivityIntent(RulesActivity.class);
        }
        return null;
    }

    private PendingIntent createActivityIntent(@NonNull Class<? extends Activity> activityClass) {
        Intent intent = new Intent(context, activityClass);
        return TaskStackBuilder.create(context)
                .addNextIntentWithParentStack(intent)
                .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    @NonNull
    private String getChannel() {
        if (Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            if (manager.getNotificationChannel(CHANNEL_ID) == null) {
                manager.createNotificationChannel(new NotificationChannel(CHANNEL_ID,
                        context.getString(R.string.notifications),
                        NotificationManager.IMPORTANCE_LOW));
            }
        }
        return CHANNEL_ID;
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/Notifications.kt
package com.bopr.android.smailer

import android.app.*
import android.app.Notification.*
import android.app.NotificationManager.IMPORTANCE_LOW
import android.content.Context
import android.content.Context.NOTIFICATION_SERVICE
import android.content.Intent
import android.os.Build
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder
import com.bopr.android.smailer.Settings.Companion.PREF_RECIPIENTS_ADDRESS
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.PREF_SENDER_ACCOUNT
import com.bopr.android.smailer.ui.*
import com.bopr.android.smailer.util.Mockable
import com.bopr.android.smailer.util.isAccountExists
import com.bopr.android.smailer.util.isValidEmailAddressList
import java.lang.System.currentTimeMillis
import kotlin.reflect.KClass

/**
 * Produces notifications.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@Mockable
class Notifications(private val context: Context) {

    private val manager = context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
    private val statusBuilder: NotificationCompat.Builder
    private val messagesBuilder: NotificationCompat.Builder
    private val errorsBuilder: NotificationCompat.Builder

    init {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            manager.createNotificationChannel(NotificationChannel(CHANNEL_ID_STATUS,
                    context.getString(R.string.status), IMPORTANCE_LOW))
            manager.createNotificationChannel(NotificationChannel(CHANNEL_ID_NOTIFICATIONS,
                    context.getString(R.string.notifications), IMPORTANCE_LOW))
        }

        statusBuilder = NotificationCompat.Builder(context, CHANNEL_ID_STATUS)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentIntent(targetIntent(TARGET_MAIN))
                .setOngoing(true)
                .apply {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        setCategory(CATEGORY_SERVICE)
                    }
                }

        messagesBuilder = NotificationCompat.Builder(context, CHANNEL_ID_NOTIFICATIONS)
                .setSmallIcon(R.drawable.ic_notification)
                .setAutoCancel(true)
                .apply {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        setCategory(CATEGORY_MESSAGE)
                    }
                }

        errorsBuilder = NotificationCompat.Builder(context, CHANNEL_ID_NOTIFICATIONS)
                .setSmallIcon(R.drawable.ic_notification)
                .setAutoCancel(true)
                .setContentText(context.getString(R.string.tap_to_check_settings))
                .apply {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        setCategory(CATEGORY_ERROR)
                    }
                }
    }

    fun serviceNotification(): Notification {
        return statusBuilder
                .setWhen(currentTimeMillis())
                .setContentTitle(context.getString(R.string.service_running))
                .build()
    }

    fun showMailSendSuccess() {
        showMessage(context.getString(R.string.email_successfully_send), null, TARGET_MAIN)
    }

    fun showRemoteAction(message: String, @Target target: Int) {
        showMessage(context.getString(R.string.remote_action), message, target)
    }

    fun showRemoteAccountError() {
        showError(REMOTE_ACCOUNT_ERROR, context.getString(R.string.service_account_not_found),
                TARGET_REMOTE_CONTROL)
    }

    fun showSenderAccountError() {
        showError(SENDER_ACCOUNT_ERROR, context.getString(R.string.sender_account_not_found),
                TARGET_MAIN)
    }

    fun showRecipientsError(@StringRes reasonRes: Int) {
        showMailError(RECIPIENTS_ERROR, reasonRes, TARGET_RECIPIENTS)
    }

    fun showGoogleAccessError() {
        showMailError(GOOGLE_ACCESS_ERROR, R.string.no_access_to_google_account, TARGET_MAIN)
    }

    internal fun cancelError(errorId: Int) {
        manager.cancel(TAG_ERROR, errorId)
    }

    internal fun onSettingsChanged(settings: Settings, key: String) {
        when (key) {
            PREF_SENDER_ACCOUNT -> {
                if (context.isAccountExists(settings.senderAccount)) {
                    cancelError(SENDER_ACCOUNT_ERROR)
                }
            }
            PREF_REMOTE_CONTROL_ACCOUNT ->
                if (context.isAccountExists(settings.remoteControlAccount)) {
                    cancelError(REMOTE_ACCOUNT_ERROR)
                }
            PREF_RECIPIENTS_ADDRESS ->
                if (isValidEmailAddressList(settings.recipients)) {
                    cancelError(RECIPIENTS_ERROR)
                }
        }
    }

    private fun showMessage(title: String, message: String? = null, @Target target: Int) {
        manager.notify(TAG_MESSAGE, nextMessageId++, messagesBuilder
                .setWhen(currentTimeMillis())
                .setContentTitle(title)
                .setContentText(message)
                .setContentIntent(targetIntent(target))
                .build())
    }

    private fun showError(errorId: Int, title: String, @Target target: Int) {
        manager.notify(TAG_ERROR, errorId, errorsBuilder
                .setWhen(currentTimeMillis())
                .setContentTitle(title)
                .setContentIntent(targetIntent(target))
                .build())
    }

    private fun showMailError(errorId: Int, @StringRes reasonRes: Int, @Target target: Int) {
        showError(errorId, context.getString(R.string.unable_send_email, context.getString(reasonRes)),
                target)
    }

    private fun targetIntent(@Target target: Int): PendingIntent {
        return when (target) {
            TARGET_MAIN ->
                activityIntent(MainActivity::class)
            TARGET_HISTORY ->
                activityIntent(HistoryActivity::class)
            TARGET_RECIPIENTS ->
                activityIntent(RecipientsActivity::class)
            TARGET_REMOTE_CONTROL ->
                activityIntent(RemoteControlActivity::class)
            TARGET_RULES ->
                activityIntent(RulesActivity::class)
            TARGET_PHONE_BLACKLIST ->
                activityIntent(CallFilterPhoneBlacklistActivity::class)
            TARGET_PHONE_WHITELIST ->
                activityIntent(CallFilterPhoneWhitelistActivity::class)
            TARGET_TEXT_BLACKLIST ->
                activityIntent(CallFilterTextBlacklistActivity::class)
            TARGET_TEXT_WHITELIST ->
                activityIntent(CallFilterTextWhitelistActivity::class)
            else ->
                throw IllegalArgumentException("Invalid target")
        }
    }

    private fun activityIntent(activityClass: KClass<out Activity>): PendingIntent {
        return TaskStackBuilder.create(context)
                .addNextIntentWithParentStack(Intent(context, activityClass.java))
                .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT)!!
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(TARGET_MAIN, TARGET_RULES, TARGET_HISTORY, TARGET_RECIPIENTS, TARGET_REMOTE_CONTROL,
            TARGET_PHONE_BLACKLIST, TARGET_PHONE_WHITELIST, TARGET_TEXT_BLACKLIST, TARGET_TEXT_WHITELIST)
    annotation class Target

    companion object {

        private var nextMessageId = 0

        private const val CHANNEL_ID_STATUS = "com.bopr.android.smailer.status"
        private const val CHANNEL_ID_NOTIFICATIONS = "com.bopr.android.smailer.notifications"

        private const val TAG_ERROR = "error"
        private const val TAG_MESSAGE = "message"

        const val SENDER_ACCOUNT_ERROR = 1000
        const val REMOTE_ACCOUNT_ERROR = 1001
        const val RECIPIENTS_ERROR = 1002
        const val GOOGLE_ACCESS_ERROR = 1003

        const val SERVICE_NOTIFICATION_ID = 19158

        const val TARGET_MAIN = 0
        const val TARGET_RULES = 1
        const val TARGET_HISTORY = 2
        const val TARGET_RECIPIENTS = 3
        const val TARGET_REMOTE_CONTROL = 4
        const val TARGET_PHONE_BLACKLIST = 5
        const val TARGET_PHONE_WHITELIST = 6
        const val TARGET_TEXT_BLACKLIST = 7
        const val TARGET_TEXT_WHITELIST = 8
    }

}